src/main/java/com/google/firebase/database/utilities/Utilities.java

```diff
@@ -18,78 +18,106 @@
-import java.net.URISyntaxException;
+import java.net.URLDecoder;
 import java.net.URLEncoder;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.Map;
+import org.apache.http.client.utils.URLEncodedUtils;
 
 public class Utilities {
-
   private static final char[] HEX_CHARACTERS = "0123456789abcdef".toCharArray();
 
   public static ParsedUrl parseUrl(String url) throws DatabaseException {
-    String original = url;
     try {
-      int schemeOffset = original.indexOf("//");
-      if (schemeOffset == -1) {
-        throw new URISyntaxException(original, "Invalid scheme specified");
-      }
-      int pathOffset = original.substring(schemeOffset + 2).indexOf("/");
-      if (pathOffset != -1) {
-        pathOffset += schemeOffset + 2;
-        String[] pathSegments = original.substring(pathOffset).split("/");
-        StringBuilder builder = new StringBuilder();
-        for (int i = 0; i < pathSegments.length; ++i) {
-          if (!pathSegments[i].equals("")) {
-            builder.append("/");
-            builder.append(URLEncoder.encode(pathSegments[i], "UTF-8"));
-          }
-        }
-        original = original.substring(0, pathOffset) + builder.toString();
-      }
+      URI uri = URI.create(url);
 
-      URI uri = new URI(original);
-      // URLEncoding a space turns it into a '+', which is different
-      // from our expected behavior. Do a manual replace to fix it.
-      final String pathString = uri.getPath().replace("+", " ");
-      Validation.validateRootPathString(pathString);
       String scheme = uri.getScheme();
+      if (scheme == null) {
+        throw new IllegalArgumentException("Database URL does not specify a URL scheme");
+      }
+
+      String host = uri.getHost();
+      if (host == null) {
+        throw new IllegalArgumentException("Database URL does not specify a valid host");
+      }
 
       RepoInfo repoInfo = new RepoInfo();
-      repoInfo.host = uri.getHost().toLowerCase();
+      repoInfo.host = host.toLowerCase();
+      repoInfo.secure = scheme.equals("https") || scheme.equals("wss");
 
       int port = uri.getPort();
       if (port != -1) {
-        repoInfo.secure = scheme.equals("https");
         repoInfo.host += ":" + port;
+      }
+
+      Map<String, String> params = getQueryParamsMap(uri.getRawQuery());
+      String namespaceParam = params.get("ns");
+      if (!Strings.isNullOrEmpty(namespaceParam)) {
+        repoInfo.namespace = namespaceParam;
       } else {
-        repoInfo.secure = true;
+        String[] parts = host.split("\\.", -1);
+        repoInfo.namespace = parts[0].toLowerCase();
       }
-      String[] parts = repoInfo.host.split("\\.");
 
-      repoInfo.namespace = parts[0].toLowerCase();
       repoInfo.internalHost = repoInfo.host;
+      // use raw (encoded) path for backwards compatibility.
+      String originalPathString = uri.getRawPath();
```

Comment:
Oh. I didn't know about this method. Awesome!

Note that you still have to turn `+` into spaces: https://github.com/firebase/firebase-android-sdk/pull/680/files#diff-76637acaa08127d536bbbe0c038845fdR72

Reply:
I don't think that we need to replace `+` with space: uri.getRawPath() doesn't seem to be outputting + instead of spaces. We have testcases to verify this.

Reply:
Yo, here are some test cases I think we should add: https://github.com/firebase/firebase-android-sdk/pull/680/files#diff-7aea264732c9130ecc3212feae077da0R53-R64
Maybe it will even show some inconsistencies -- let's see.

Reply:
I added a test case for `+` being converted to ` `, but the non-ASCII characters testcase blew up the Java URI class.